Source/Core/VideoCommon/TextureConverterShaderGen.cpp

```diff
@@ -0,0 +1,247 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <array>
+#include <cstring>
+
+#include "Common/Assert.h"
+#include "Common/CommonTypes.h"
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/TextureConverterShaderGen.h"
+#include "VideoCommon/VideoCommon.h"
+#include "VideoCommon/VideoConfig.h"
+
+TextureConverterShaderUid GetTextureConverterShaderUid(EFBCopyFormat dst_format, bool is_depth_copy,
+                                                       bool is_intensity, bool scale_by_half)
+{
+  TextureConverterShaderUid out;
+  convertion_shader_uid_data* uid_data = out.GetUidData<convertion_shader_uid_data>();
+  memset(uid_data, 0, sizeof(*uid_data));
+
+  uid_data->dst_format = dst_format;
+  uid_data->efb_has_alpha = bpmem.zcontrol.pixel_format == PEControl::RGBA6_Z24;
+  uid_data->is_depth_copy = is_depth_copy;
+  uid_data->is_intensity = is_intensity;
+  uid_data->scale_by_half = scale_by_half;
+
+  return out;
+}
+
+ShaderCode GenerateTextureConverterShaderCode(APIType api_type,
+                                              const convertion_shader_uid_data* uid_data)
+{
+  ShaderCode out;
+
+  out.Write(
+      "SAMPLER_BINDING(9) uniform sampler2DArray samp9;\n"
+      "in vec3 f_uv0;\n"
+      "out vec4 ocol0;\n"
+      "void main(){\n"
+      "	vec4 texcol = texture(samp9, %s);\n",
+      uid_data->is_depth_copy ?
+          (g_ActiveConfig.bStereoEFBMonoDepth ? "vec3(f_uv0.xy, 0.0)" : "vec3(f_uv0.xy, f_uv0.z)") :
+          "f_uv0");
+
+  if (uid_data->is_depth_copy)
+  {
+    out.Write("	int depth = int(texcol.x * 16777216.0);\n"
```

Comment:
No, Dolphin uses a power-of-two factor to prevent loss of precision. This is very important for Dolphin because we use the full 23 decimal bits, so anything that is not simply a shift of the 8-bit exponent will cause issues.

We enable depth clamping to make use the depth values are always clamped to the 0..2^24-1 depth range.